src/backtrace.rs

```diff
@@ -0,0 +1,77 @@
+// SPDX-FileCopyrightText: 2023 Rivos Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+/// Get an array of backtrace addresses.
+///
+/// This needs `force-frame-pointers` enabled.
+use core::arch::asm;
+
+pub const MAX_BACKTRACE_ADDRESSES: usize = 10;
+
+extern "C" {
+    static _stack_start: u8;
+    static _stack_end: u8;
+}
+
+#[derive(Copy, Clone)]
+pub(crate) enum BTReturn {
+    Some(usize),
+    None,
+    BadFp(usize),
+    BadAddress(usize),
+}
+
+pub(crate) fn backtrace() -> [BTReturn; MAX_BACKTRACE_ADDRESSES] {
+    // Safe because we are just reading a register
+    let fp = unsafe {
+        let mut tmp: u64;
+        asm!("mv {0}, x8", out(reg) tmp);
+        tmp
+    };
+
+    unsafe { backtrace_internal(fp) }
```

Comment:
Needs a comment explaining why it's safe, and also corresponding documentation on the caller requirements in `backtrace_internal()`. 

Probably something like:

```
// # Safety
// The caller must ensure that fp is a valid address on the stack frame.
unsafe fn backtrace_internal(fp: u64) -> [BTReturn; MAX_BACKTRACE_ADDRESSES] {
```

Reply:
fixed